File: src/data/courses/dsa/binary-search-on-answer/distance-maximization/magnetic-force-balls/code/java/MagneticForceBalls.java
class Solution {
    public int maxDistance(int[] pos, int m) {
        Arrays.sort(pos); int lo=1,hi=pos[pos.length-1]-pos[0];
        while(lo<hi){int mid=lo+(hi-lo+1)/2,cnt=1,last=pos[0]; for(int i=1;i<pos.length;i++) if(pos[i]-last>=mid){cnt++;last=pos[i];} if(cnt>=m)lo=mid;else hi=mid-1;}
        return lo;
    }
}

File: src/data/courses/dsa/binary-search-on-answer/distance-maximization/magnetic-force-balls/code/python/magnetic_force_balls.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        lo, hi = 1, position[-1] - position[0]
        while lo < hi:
            mid = lo + (hi - lo + 1) // 2
            count, last = 1, position[0]
            for p in position[1:]:
                if p - last >= mid: count += 1; last = p
            if count >= m: lo = mid
            else: hi = mid - 1
        return lo